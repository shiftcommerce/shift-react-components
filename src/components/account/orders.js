// Libraries
import React, { Component } from 'react'

// Lib
import componentMapping from '../../lib/component-mapping'

class AccountOrders extends Component {
  constructor (props) {
    super(props)

    this.Button = componentMapping('Button')
    this.Link = componentMapping('Link')
    this.Loading = componentMapping('Loading')
    this.OrderList = componentMapping('OrderList')
  }

  componentDidMount () {
    this.props.fetchOrders()
  }

  renderOrdersList (orders) {
    if (orders.data.length === 0) {
      return (<p>No previous orders found.</p>)
    }

    return (<this.OrderList orders={orders} />)
  }

  renderAccountBanner () {
    return (
      <div className='c-order-history__nav'>
        <h2>Order History</h2>
      </div>
    )
  }

  render () {
    const { orders, orders: { loading } } = this.props

    if (loading) {
      return (
        <>
          <div className='c-order-history'>
            { this.renderAccountBanner() }
            <this.Loading />
          </div>
        </>
      )
    } else {
      return (
        <div className='c-order-history'>
          { this.renderAccountBanner() }
          { this.renderOrdersList(orders) }
        </div>
      )
    }
  }
}

export default AccountOrders
