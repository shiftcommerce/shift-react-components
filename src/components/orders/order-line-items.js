// Libraries
import React, { PureComponent } from 'react'
import t from 'typy'

// Lib
import { decimalPrice } from '../../lib/decimal-price'
import { penceToPounds } from '../../lib/pence-to-pounds'
import componentMapping from '../../lib/component-mapping'

class OrderLineItems extends PureComponent {
  constructor (props) {
    super(props)

    this.Image = componentMapping('Image')
    this.Link = componentMapping('Link')
  }

  /**
  * Render the total prices for the line item
  * @param  {Object} lineItem
  * @return {string} - HTML markup for the component
  */
  renderTotal (lineItem) {
    const total = t(lineItem, 'pricing.line_total_inc_tax').safeObject
    const formattedTotal = decimalPrice(total)

    return (
      <div className='c-order-history__line-items_grid_item2'>
        <a>Â£{ penceToPounds(formattedTotal) }</a>
      </div>
    )
  }

  /**
  * Render the title & sku for the line item
  * @param  {Object} lineItem
  * @return {string} - HTML markup for the component
  */
  renderTitle (lineItem) {
    // To be updated, see: https://github.com/shiftcommerce/shift-front-end-react/issues/713
    return (
      <>
        <h5>{ lineItem.sku }</h5>
        <h4>{ lineItem.sku }</h4>
      </>
    )
  }

  /**
  * Render the quantity, delivery method & customer name for the line item
  * @param  {Object} lineItem
  * @return {string} - HTML markup for the component
  */
  renderParams (lineItem) {
    const deliveryMethod = t(lineItem, 'shipping_method.label').safeObject
    const customerName = t(lineItem, 'shipping_address.name').safeObject

    return (
      <div className='c-order-history__line-items_grid_item1'>
        { this.renderTitle(lineItem) }
        <span>Quantity: { lineItem.quantity }</span>
        <span>Delivery: { deliveryMethod }</span>
        <span>Address: { customerName }</span>
      </div>
    )
  }

  renderLineItems () {
    const { items } = this.props

    return (
      items.map((lineItem) =>
        <div className='c-order-history__line-items_grid' key={lineItem.sku}>
          { this.renderParams(lineItem) }
          { this.renderTotal(lineItem) }
        </div>
      )
    )
  }

  render () {
    return this.renderLineItems()
  }
}

export default OrderLineItems
