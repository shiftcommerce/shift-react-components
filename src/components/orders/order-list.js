// Libraries
import React, { PureComponent } from 'react'
import dateFns from 'date-fns'
import t from 'typy'

// Lib
import { penceToPounds } from '../../lib/pence-to-pounds'
import componentMapping from '../../lib/component-mapping'

class OrderList extends PureComponent {
  constructor (props) {
    super(props)

    this.OrderLineItems = componentMapping('OrderLineItems')
    this.ShippingAddresses = componentMapping('ShippingAddresses')
  }

  /**
  * Render the total shipping costs in the order list
  * @param  {Object} order
  * @return {string} - HTML markup for the component
  */
  renderShippingTotal (order) {
    const shippingTotal = order.shipping_methods.reduce(
      (accumulator, currentValue) => accumulator + currentValue.price, 0
    )

    return penceToPounds(shippingTotal)
  }

  /**
  * Render the total prices for the line item
  * @param  {Object} order
  * @param  {Object} orderDate
  * @param  {Object} total
  * @return {string} - HTML markup for the component
  */
  renderOrderSummary (order, orderDate, total) {
    return (
      <div className='c-order-history__order-summary'>
        <div className='c-order-history__order-summary_params'>
          <p className='u-bold u-inline-block'>Date of Order: </p>
          <a className='u-inline-block'>{ orderDate }</a>
        </div>
        <div className='c-order-history__order-summary_params'>
          <p className='u-bold u-inline-block'>Items: </p>
          <a className='u-inline-block'>{ order.line_items.length }</a>
        </div>
        <div className='c-order-history__order-summary_params'>
          <p className='u-bold u-inline-block'>Order Number: </p>
          <a className='u-inline-block'>{ order.reference }</a>
        </div>
        <div className='c-order-history__order-summary_params'>
          <p className='u-bold u-inline-block'>Total: </p>
          <a className='u-inline-block'>&pound;{ total }</a>
        </div>
      </div>
    )
  }

  /**
  * Render the order details
  * @param  {Object} order
  * @param  {Object} orderDate
  * @param  {Object} total
  * @return {string} - HTML markup for the component
  */
  renderOrderDetails (order, orderDate, total) {
    return (
      <div className='c-order-history__details'>
        { this.renderOrderSummary(order, orderDate, total) }
        <this.OrderLineItems items={order.line_items} />
        <div className='c-order-history__totals'>
          <p>Shipping: &pound;{ this.renderShippingTotal(order) } </p>
          <p className='u-bold'>Total: &pound;{ total }</p>
        </div>
        <h3>Shipping Details</h3>
        <this.ShippingAddresses addresses={order.shipping_addresses} />
      </div>
    )
  }

  /**
  * Render the header for the order list
  * @param  {Object} order
  * @param  {Object} orderDate
  * @param  {Object} total
  * @return {string} - HTML markup for the component
  */
  renderOrderHeader (order, orderDate, total) {
    return (
      <>
        <h4>{ orderDate } - { order.reference } - Â£{ total }</h4>
        <label htmlFor={order.reference} className='c-order-history__header-label' />
        <input type='checkbox' id={order.reference} />
        <span className='c-order-history__arrow' />
      </>
    )
  }

  render () {
    const { orders } = this.props

    return (
      <>
        {orders.data.map((order) => {
          const orderDate = dateFns.format(new Date(order.placed_at), 'MMM D, YYYY')
          const total = penceToPounds(t(order, 'pricing.total_inc_tax').safeObject)

          return (
            <div className='c-order-history__header' key={order.id}>
              { this.renderOrderHeader(order, orderDate, total) }
              { this.renderOrderDetails(order, orderDate, total) }
            </div>
          )
        })}
      </>
    )
  }
}

export default OrderList
