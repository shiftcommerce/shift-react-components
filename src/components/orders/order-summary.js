// Libraries
import React, { Component } from 'react'
import Sticky from 'react-stickyfill'

// Lib
import { decimalPrice } from '../../lib/decimal-price'
import componentMapping from '../../lib/component-mapping'

class OrderSummary extends Component {
  constructor (props) {
    super(props)

    this.Image = componentMapping('Image')
    this.Link = componentMapping('Link')
  }

  /**
  * Render line item image for the order summary
  * @param  {Object} lineItem
  * @return {string} - HTML markup for the component
  */
  renderLineItemImage (lineItem) {
    if (lineItem.imageUrl) {
      return (
        <this.Link href={`/slug?slug=${lineItem.slug}`} as={`${lineItem.canonical_path}`} >
          <this.Image src={lineItem.imageUrl} alt={lineItem.title} aria-label={lineItem.title} />
        </this.Link>
      )
    } else {
      return (
        <this.Image src='/static/placeholder.png' alt={lineItem.title} aria-label={lineItem.title} />
      )
    }
  }

  /**
  * Render the line item title for the order summary
  * @param  {Object} lineItem
  * @return {string} - HTML markup for the component
  */
  renderLineItemTitle (lineItem) {
    return (
      <>
        <p className='c-order__summary-line-item-details-title u-bold'>
          { lineItem.title }
        </p>
        <p className='c-order__summary-line-item-details-sku'>
          { lineItem.sku }
        </p>
      </>
    )
  }

  /**
  * Render the line-item meta-data for the order summary
  * @param  {Object} _lineItem
  * @return {string} - HTML markup for the component
  */
  renderLineItemParams (_lineItem) {
    // @TODO: to update this with actual metadata
    const metaData = { colour: 'Grey', size: 'L', quantity: 1 }
    return (
      Object.keys(metaData).map((key, index) =>
        <div className='c-order__summary-line-item-details-param' key={key}>
          <div className='c-order__summary-line-item-details-param-key'>
            <p>{ key }:</p>
          </div>
          <div className='c-order__summary-line-item-details-param-value'>
            <p>{ metaData[key] }</p>
          </div>
        </div>
      )
    )
  }

  /**
  * Render the line items for the order summary
  * @param  {Object} order
  * @return {string} - HTML markup for the component
  */
  renderLineItems (order) {
    return (
      order.line_items.map((lineItem) =>
        <div className='c-order__summary-line-item' key={lineItem.sku}>
          <div className='c-order__summary-line-item-image'>
            { this.renderLineItemImage(lineItem) }
          </div>

          <div className='c-order__summary-line-item-details'>
            { this.renderLineItemTitle(lineItem) }
            { this.renderLineItemParams(lineItem) }
          </div>
        </div>
      )
    )
  }

  /**
  * Render the total prices for the order summary
  * @param  {Object} order
  * @return {string} - HTML markup for the component
  */
  renderOrderTotals (order) {
    return (
      <Sticky>
        <div aria-label='Order total summary' className='u-sticky'>
          <div>
            <dl aria-label='Order subtotal'>
              <dt>Total products:</dt>
              <dd>&pound;{ decimalPrice(order.sub_total) }</dd>
            </dl>
            <dl aria-label='Order shipping costs'>
              <dt>Shipping costs:</dt>
              <dd>&pound;{ order.shipping_total }</dd>
            </dl>
            <dl aria-label='Order total' className='u-bold'>
              <dt>TOTAL:</dt>
              <dd>&pound;{ decimalPrice(order.total) }</dd>
            </dl>
            <dl>
              <dt>* Including VAT</dt>
            </dl>
          </div>
        </div>
      </Sticky>
    )
  }

  render () {
    const { order } = this.props

    return (
      <div className='c-order__summary'>
        <div className='c-order__summary-header'>
          <div className='c-order__summary-header-title'>
            <h2>Your Order</h2>
          </div>

          <div className='c-order__summary-header-total'>
            <h2>&pound;{ decimalPrice(order.total) }</h2>
          </div>
        </div>

        { this.renderLineItems(order) }

        <div className='c-order__summary-totals'>
          { this.renderOrderTotals(order) }
        </div>
      </div>
    )
  }
}

export default OrderSummary
