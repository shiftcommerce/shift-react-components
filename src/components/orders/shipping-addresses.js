// Libraries
import React, { PureComponent } from 'react'

class ShippingAddresses extends PureComponent {
  render () {
    return (
      <>
        { this.props.addresses.map((shippingAddress) => {
          return (
            <div className='c-order-history__shipping-address' key={shippingAddress.id}>
              <p className='u-bold'>{ shippingAddress.name }</p>
              <p>{ shippingAddress.company }</p>
              <p>{ shippingAddress.lines.join() }</p>
              <p>{ shippingAddress.city }</p>
              <p>{ shippingAddress.state }</p>
              <p>{ shippingAddress.postcode }</p>
              <p>{ shippingAddress.country }</p>
            </div>
          )
        })}
      </>
    )
  }
}

export default ShippingAddresses
