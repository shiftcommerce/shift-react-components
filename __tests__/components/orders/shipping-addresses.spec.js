// Libraries
import React from 'react'

// Component
import ShippingAddresses from '../../../src/components/orders/shipping-addresses'

// Fixtures
import orders from '../../fixtures/orders'

test('renders correctly', () => {
  // Arrange
  const shippingAddresses = orders.data[0].shipping_addresses

  // Act
  const wrapper = mount(
    <ShippingAddresses addresses={shippingAddresses} />
  )

  // Assert
  expect(wrapper).toMatchSnapshot()
  expect(wrapper).toIncludeText(shippingAddresses[0].name)
  expect(wrapper).toIncludeText(shippingAddresses[0].company)
})
