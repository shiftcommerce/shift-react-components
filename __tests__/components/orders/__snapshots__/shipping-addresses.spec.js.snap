// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<ShippingAddresses
  addresses={
    Array [
      Object {
        "city": "Leeds",
        "company": "SHIFT Commerce",
        "country": "United Kingdom",
        "id": "61c40964-4b4c-4f54-b580-958798087312_1",
        "lines": Array [
          "The Calls",
        ],
        "meta_attributes": Object {
          "further": "information",
        },
        "name": "Test Customer",
        "postcode": "LS2 7EY",
        "state": "West Yorkshire",
      },
      Object {
        "city": "Leeds",
        "company": null,
        "country": "United Kingdom",
        "id": "61c40964-4b4c-4f54-b580-958798087312_2",
        "lines": Array [
          "SHIFT Commerce",
          "The Calls Landing",
        ],
        "meta_attributes": Object {
          "further": "information",
        },
        "name": "Test Customer 2",
        "postcode": "LS2 7EY",
        "state": "West Yorkshire",
      },
    ]
  }
>
  <div
    className="c-order-history__shipping-address"
    key="61c40964-4b4c-4f54-b580-958798087312_1"
  >
    <p
      className="u-bold"
    >
      Test Customer
    </p>
    <p>
      SHIFT Commerce
    </p>
    <p>
      The Calls
    </p>
    <p>
      Leeds
    </p>
    <p>
      West Yorkshire
    </p>
    <p>
      LS2 7EY
    </p>
    <p>
      United Kingdom
    </p>
  </div>
  <div
    className="c-order-history__shipping-address"
    key="61c40964-4b4c-4f54-b580-958798087312_2"
  >
    <p
      className="u-bold"
    >
      Test Customer 2
    </p>
    <p />
    <p>
      SHIFT Commerce,The Calls Landing
    </p>
    <p>
      Leeds
    </p>
    <p>
      West Yorkshire
    </p>
    <p>
      LS2 7EY
    </p>
    <p>
      United Kingdom
    </p>
  </div>
</ShippingAddresses>
`;
